File: v1/kotlin/pascals-triangle/src/main/kotlin/PascalsTriangle.kt
object PascalsTriangle {
    fun row(i: Int): List<Int> {
        val values = mutableListOf(1)
        var column = 1

        for (j in 1 .. i - 1) {
            column = column * (i - j) / j
            values.add(column)
        }

        return values;
    }

    fun computeTriangle(rows: Int): List<List<Int>> =
        if (rows < 0) throw IllegalArgumentException()
        else (1..rows).map { row(it) }
}

File: v1/ecmascript/pascals-triangle/pascals-triangle.js
export default class Triangle {
    constructor(size) {
        this.size = size;
    }

    get rows() {
        let acc = [];

        for (let i = 1; i <= this.size; i++)
            acc.push(this.row(i));

        return acc;
    }

    get lastRow() {
        return this.row(this.size);
    }

    row(n) {
        let digits = [1];
        let column = 1;

        for (let j = 1; j < n; j++) {
            column = column * (n - j) / j;
            digits.push(column);
        }

        return digits;
    }
}